### PR TITLE
CNN and RNN

### PR DESCRIPTION
**Pull request by reveurmichael**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] [local build the book](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-book)
- [x] review the artifacts from local build
- [ ] [local build the slides](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-slides-optional) if appliable
- [ ] Unit test if appliable

# Checklist:

- [ ] My code follows the [style guidelines](https://github.com/open-academy/machine-learning/blob/main/STYLE_GUIDE.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have applied tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: open-machine-learning-jupyter-book/assets/html-iframes/conv-demo/index.html [hunk 1]
```diff
@@ -0,0 +1,326 @@
+
```
[the remark this answers is not in the document]
Comment (zhangqi444) [reply open-machine-learning-jupyter-book/assets/html-iframes/conv-demo/index.html:1]: Where is such a file from? Could we have the copyright info added? Even to MIT, we need to declare the author info. You can copy the original license file to `conv-demo/` folder

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 2]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
   name: python3
 ---
 
-# CNN (TBD)
\ No newline at end of file
+
+# Convolutional Neural Networks
+
+
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:19]: remove extra empty line

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 3]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
...
+
+
+```{epigraph}
+Thanks to Convolutional Neural Network, computer vision is working far better than just two years ago, and this is enabling numerous exciting applications ranging from safe autonomous driving, to accurate face recognition, to automatic reading of radiology images.
+
+-- Andrew Ng
+```
+
+
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:27]: remove, and apply to all the other places

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 4]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
...
+:::{figure-md} 01_intro_cnn-dl
+<img src="../../images/deep-learning/CNN/01_intro_cnn.png" width="90%" class="bg-white mb-1">
+
+Illustration of matrix mutliplication (filter) in CNN {cite}`reluwiki`
+:::
+
+
+
+<iframe src="../html-iframes/conv-demo/index.html" width="105%" height="700px;" style="border:none;"></iframe>
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:43]: Do you want to give any description to this iframe section? As I can't see the screenshot, I don't know what this is for. E.g. you can include it to a ```{seealso} section if it is a further reading staff.

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 5]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
...
+:::{figure-md} 02_cnn1_mnist_output-dl
+<img src="../../images/deep-learning/CNN/02_cnn1_mnist_output.png" width="90%" class="bg-white mb-1">
+
+Train MINIST dataset with CNN: prediction output
+:::
+
+
+A random set of 6 digits with actual and predicted labels. You can see a prediction failure in the lower right box.
+
+### Code
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:91]: Can we break down this big code section to multiple {code-cell}, and add some explanation correspondingly? It is very difficult to follow such a long code in a book style doc.

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 6]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
...
+
+
+## Stylenet / Neural-Style
+The purpose of this script is to illustrate how to do stylenet in TensorFlow. We reference the following paper(link: https://arxiv.org/abs/1508.06576) for this algorithm.
+
+But there is some prerequisites
+- Download the VGG-verydeep-19.mat file.
+- You must download two images, a style image and a content image for the algorithm to blend.
+
+The style image is
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:744]: The style image is like below. 

Apply the similar tough to all below.

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 7]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
...
+# Create labels file
+cifar_labels_file = os.path.join(data_dir,'cifar10_labels.txt')
+print('Writing labels file, {}'.format(cifar_labels_file))
+with open(cifar_labels_file, 'w') as labels_file:
+    for item in objects:
+        labels_file.write("{}\n".format(item))
+```
+
+
+## Stylenet / Neural-Style
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:737]: neural-style? We only uppercase the first letter.

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 8]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
...
+The final result looks like
+
+
+:::{figure-md} 05_stylenet_ex-dl
+<img src="../../images/deep-learning/CNN/05_stylenet_ex.png" width="90%" class="bg-white mb-1">
+
+stylenet final result
+:::
+
+### Code
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:771]: same as above

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 9]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
...
+    showarray(img0/255.0)
+
+    # Create deep dream
+    render_deepdream(T(layer)[:, :, :, channel], img0, iter_n=15)
+
+    sess.close()
+```
+
+
+---
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:1183]: remove

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 10]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
...
+
+## Your turn! 🚀
+
+Practice the Convolutional Neural Networks by following this xxxx.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:1187]: xxxx -> TBD

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 11]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
...
+
+## Self study
+
+You can refer to those YouTube videos for further study:
+- [Convolutional Neural Networks (CNNs) explained, by deeplizard](https://www.youtube.com/watch?v=YRhxdVk_sIs)
+- [Convolutional Neural Networks Explained (CNN Visualized), by Futurology](https://www.youtube.com/watch?v=pj9-rr1wDhM)
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:1193]: use iframe?

File: open-machine-learning-jupyter-book/deep-learning/CNN.md [hunk 12]
```diff
@@ -13,4 +13,1201 @@ kernelspec:
...
+
+
+## [optional] At the frontier of the research world
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/CNN.md:1196]: We don't have such section. You could make it as 3rd level header as part of self study

File: open-machine-learning-jupyter-book/deep-learning/RNN.md [hunk 13]
```diff
@@ -13,9 +13,292 @@ kernelspec:
...
+```{epigraph}
+The nervous system contains many circular paths, whose activity so regenerates the excitation of any participant neuron that reference to time past becomes indefinite, although it still implies that afferent activity has realized one of a certain class of configurations over time. Precise specification of these implications by means of recursive functions, and determination of those that can be embodied in the activity of nervous nets, completes the theory.
+
+-- Warren McCulloch and Walter Pitts, 1943
+```
+
+
+A recurrent neural network (RNN) is a type of artificial neural network which uses sequential data or time series data and it is mainly used for Natural Language Processing. Now let us see what it looks like.
+
+Sequential data is not i.i.d.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/RNN.md:28]: what's i.i.d? Make it a link or add a ref.

File: open-machine-learning-jupyter-book/deep-learning/RNN.md [hunk 14]
```diff
@@ -13,9 +13,292 @@ kernelspec:
...
+plt.plot(epoch_seq, test_accuracy, 'r-', label='Test Set')
+plt.title('Test Accuracy')
+plt.xlabel('Epochs')
+plt.ylabel('Accuracy')
+plt.legend(loc='upper left')
+plt.show()
+```
+
+
+---
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/RNN.md:283]: remove

File: open-machine-learning-jupyter-book/deep-learning/RNN.md [hunk 15]
```diff
@@ -13,9 +13,292 @@ kernelspec:
...
+
+## Your turn! 🚀
+
+Practice the Recurrent Neural Networks by following this xxxx.
+
+## Self study
+
+
+## [optional] At the frontier of the research world
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/RNN.md:292]: If you dont have any content, use TBD

File: open-machine-learning-jupyter-book/deep-learning/RNN.md [hunk 16]
```diff
@@ -13,9 +13,292 @@ kernelspec:
...
+plt.legend(loc='upper left')
+plt.show()
+```
+
+
+---
+
+## Your turn! 🚀
+
+Practice the Recurrent Neural Networks by following this xxxx.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/RNN.md:287]: xxxx -> TBD

File: open-machine-learning-jupyter-book/deep-learning/CNN.md (hunk 12 above)
Comment (zhangqi444) [reply open-machine-learning-jupyter-book/deep-learning/CNN.md:1196]: You can use 'Research trend' as the title? Or any other better name.

**Review by zhangqi444 — APPROVED** [review]
In future, please try to have one PR for one section. We should ALWAYS try to have smaller PRs more frequently.

File: open-machine-learning-jupyter-book/assets/html-iframes/conv-demo/external/d3.min.js [hunk 17]
```diff
@@ -0,0 +1,5 @@
...e)}function u(){a.point=t}function i(){n.closePath()}var o=4.5,a={point:t,lineStart:function(){a.point=e},lineEnd:u,polygonStart:function(){a.lineEnd=i},polygonEnd:function(){a.lineEnd=u,a.point=t},pointRadius:function(n){return o=n,a},result:y};return a}function nr(n){function t(n){return(a?r:e)(n)}function e(t){return rr(t,function(e,r){e=n(e,r),t.point(e[0],e[1])})}function r(t){function e(e,r){e=n(e,r),t.point(e[0],e[1])}function r(){M=0/0,S.point=i,t.lineStart()}function i(e,r){var i=pe([e,r]),o=n(e,r);u(M,x,y,b,_,w,M=o[0],x=o[1],y=e,b=i[0],_=i[1],w=i[2],a,t),t.point(M,x)}function o(){S.point=e,t.lineEnd()}function c(){r(),S.point=l,S.lineEnd=s}function l(n,t){i(f=n,h=t),g=M,p=x,v=b,d=_,m=w,S.point=i}function s(){u(M,x,y,b,_,w,g,p,f,v,d,m,a,t),S.lineEnd=o,o()}var f,h,g,p,v,d,m,y,M,x,b,_,w,S={point:e,lineStart:r,lineEnd:o,polygonStart:function(){t.polygonStart(),S.lineStart=c},polygonEnd:function(){t.polygonEnd(),S.lineStart=r}};return S}function u(t,e,r,a,c,l,s,f,h,g,p,v,d,m){var y=s-t,M=f-e,x=y*y+M*M;if(x>4*i&&d--){var b=a+g,_=c+p,w=l+v,S=Math.sqrt(b*b+_*_+w*w),k=Math.asin(w/=S),E=va(va(w)-1)<Ta||va(r-h)<Ta?(r+h)/2:Math.atan2(_,b),A=n(E,k),N=A[0],C=A[1],z=N-t,q=C-e,L=M*z-y*q;
```
[the remark this answers is not in the document]
Comment (reveurmichael) [reply open-machine-learning-jupyter-book/assets/html-iframes/conv-demo/external/d3.min.js:1]: Should be alright with your solution. Will try

File: open-machine-learning-jupyter-book/assets/html-iframes/conv-demo/index.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (reveurmichael) [reply open-machine-learning-jupyter-book/assets/html-iframes/conv-demo/index.html:1]: OK, will do that.

File: open-machine-learning-jupyter-book/deep-learning/RNN.md (hunk 15 above)
Comment (reveurmichael) [inline open-machine-learning-jupyter-book/deep-learning/RNN.md:292]: Kaixin, please help remove this [optional] section. Thank you!

File: STYLE_GUIDE.md [hunk 18]
```diff
@@ -55,11 +55,24 @@ Use warm, inclusive language (such as “them”, even when referring to a singl
...
+    * White as background (text is black)
       * (Try to avoid) Extra Dark：3FA3FD
       * Dark：66BFFF
       * Light：B2D9FF
       * (Try to avoid) Extra Light: CFF4FF
+      
+## Figure in MarkDown
+You are recommended to use markdown for adding a figure. The corresponding markdown code is:
+
+```text
```
[the remark this answers is not in the document]
Comment (reveurmichael) [reply STYLE_GUIDE.md:67]: Done

File: STYLE_GUIDE.md [hunk 19]
```diff
@@ -55,11 +55,24 @@ Use warm, inclusive language (such as “them”, even when referring to a singl
...
+:::{figure-md} figure_label
+<img src="path/to/your/figure/file" width="90%" class="bg-white mb-1">
+
+Caption for the Figure (markdown is supported for hyperlinks, references, etc.)
+:::
+```
+
+Typically, we set the width to `90%`. However, for some smaller figures, you could change that to other values such as `50%` or `30%`.
```
[the remark this answers is not in the document]
Comment (reveurmichael) [reply STYLE_GUIDE.md:75]: It's not necessary. No impact.

File: STYLE_GUIDE.md [hunk 20]
```diff
@@ -55,11 +55,24 @@ Use warm, inclusive language (such as “them”, even when referring to a singl
...
     * Do not italicize numbers or parentheses in subscripts or superscripts
   * Color：
-    * Blue as background (text is black)
+    * White as background (text is black)
       * (Try to avoid) Extra Dark：3FA3FD
       * Dark：66BFFF
       * Light：B2D9FF
       * (Try to avoid) Extra Light: CFF4FF
+      
+## Figure in MarkDown
```
[the remark this answers is not in the document]
Comment (reveurmichael) [reply STYLE_GUIDE.md:64]: Done

File: open-machine-learning-jupyter-book/deep-learning/RNN.md (hunk 15 above)
Comment (reveurmichael) [reply open-machine-learning-jupyter-book/deep-learning/RNN.md:292]: Done

File: open-machine-learning-jupyter-book/assets/html-iframes/conv-demo/index.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (reveurmichael) [reply open-machine-learning-jupyter-book/assets/html-iframes/conv-demo/index.html:1]: Done
